test/compatibility_lifecycle/cmd/feature_gates.go

```diff
@@ -205,27 +199,70 @@ func dedupeFeatureList(featureList []featureInfo) ([]featureInfo, error) {
 	return deduped, nil
 }
 
-func verifyFeatureDeletionOnly(newFeatureList []featureInfo, oldFeatureList []featureInfo) error {
-	oldFeatureSet := make(map[string]*featureInfo)
-	for _, f := range oldFeatureList {
-		oldFeatureSet[f.Name] = &f
-	}
-	newFeatures := []string{}
-	for _, f := range newFeatureList {
-		oldSpecs, found := oldFeatureSet[f.Name]
-		if !found {
-			newFeatures = append(newFeatures, f.Name)
-		} else if !reflect.DeepEqual(*oldSpecs, f) {
-			return fmt.Errorf("feature %s changed with diff: %s", f.Name, cmp.Diff(*oldSpecs, f))
+// verifyFeatureRemoval checks if removed features are allowed to be removed based on their lifecycle.
+// Alpha features can be removed anytime without error.
+// Returns error if:
+//   - Beta features are removed (not allowed)
+//   - GA/Deprecated features are removed without being locked to default
+//   - GA/Deprecated features locked after v1.31 are removed before 3 minor versions
+//     have passed (required for emulation support)
+func verifyFeatureRemoval(featureList []featureInfo, baseFeatureList []featureInfo,
+	currentVersion *version.Version, thresholdVersion *version.Version) error {
+	if thresholdVersion == nil {
+		thresholdVersion = version.MajorMinor(0, 0)
+	}
+	baseFeatures := make(map[string]featureInfo)
+	for _, f := range baseFeatureList {
+		baseFeatures[f.Name] = f
+	}
+	currentFeatures := make(map[string]featureInfo)
+	for _, f := range featureList {
+		currentFeatures[f.Name] = f
+	}
+
+	for name, baseFeature := range baseFeatures {
+		// Check if feature was removed
+		if _, found := currentFeatures[name]; found {
+			continue
+		}
+
+		// Feature was removed, check if allowed
+		specs := baseFeature.VersionedSpecs
+		if len(specs) == 0 {
+			return fmt.Errorf("feature %s has no version specs", name)
+		}
+
+		lastSpec := specs[len(specs)-1]
+		switch lastSpec.PreRelease {
+		case "Alpha":
+			continue // can remove alpha features anytime
+		case "Beta":
+			return fmt.Errorf("feature %s cannot be removed while in beta", name)
+		case "GA", "Deprecated":
+			if !lastSpec.LockToDefault {
+				return fmt.Errorf("feature %s cannot be removed because it is in GA or Deprecated state and is not locked to default", name)
+			}
+			specVer, err := version.Parse(lastSpec.Version)
+			if err != nil {
+				return fmt.Errorf("invalid version \"%s\" for feature %s: %w", lastSpec.Version, name, err)
+			}
+			// we do not require the 3 version retention for features locked before the thresholdVersion.
+			// TODO: remove after 1.34
+			if !specVer.GreaterThan(thresholdVersion) {
+				continue
+			}
+			minRemovalVer := specVer.AddMinor(3)
```

Comment:
we might want to pull this constant out somewhere more visible or even public, for if/when this number is revisited.

NOT blocking, just a thought